asdf/block.py

```diff
@@ -554,20 +555,31 @@ def get_external_uri(self, uri, index):
         parts[2] = path
         return patched_urllib_parse.urlunparse(parts)
 
-    def _find_used_blocks(self, tree, ctx):
+    def _find_used_blocks(self, tree, ctx, remove=True):
         reserved_blocks = set()
 
         for node in treeutil.iter_tree(tree):
-            # check that this object will not be handled by a converter
-            if not ctx.extension_manager.handles_type(type(node)):
+            if ctx.extension_manager.handles_type(type(node)):
+                converter = ctx.extension_manager.get_converter_for_type(type(node))
+                sctx = ctx._create_serialization_context()
+                tag = converter.select_tag(node, sctx)
+                for key in converter.reserve_blocks(node, tag, sctx):
```

Comment:
The original thought was that different tag versions might require different block usage. I guess one option might be to pass in the tag but not the serialization context but this felt at odds with the other parts of the Converter API. The serialization context would still need to be provided to `select_tag` so it's possible a converter could mess with blocks at that point as well.

One solution to this might be a more aware serialization context. Something that knows if this is a validation, a write, a read, etc and only allows modifications that are consistent with that operation (e.g. on validate, don't modify, allocate or free blocks). I made a brief attempt at something like this but ran into issues with the block manager (it's difficult to interact without producing a change in it's state).

So some options to address this might be:
- drop the serialization context (and possibly tag) argument to `reserve_blocks`
- leave in the argument(s) and add tests to cover their usage

As there is no current use of the serialization context in `reserve_blocks` I'm not sure what test should be written to cover it's usage. Are there examples of where this argument is used in other converters?

I'm good with either option at this point. I'm also not seeing `SerializationContext` listed in our user or developer api documentation but attempting to add a fix for that to this PR seems like scope creep.

Reply:
Dropped the `ctx` arg here: 93c2eac16784bf490fee5e7851d94bdbc1509dd0